concordium-std/src/state_btree.rs

```diff
@@ -0,0 +1,1481 @@
+    pub fn len(&self) -> u32 { self.len }
+
+    /// Returns `true` is the set contains no elements.
+    pub fn is_empty(&self) -> bool { self.root.is_none() }
+
+    /// Get an iterator over the elements in the `StateBTreeSet`. The iterator
+    /// returns elements in increasing order.
+    pub fn iter(&self) -> StateBTreeSetIter<K, S, M>
+    where
+        S: HasStateApi, {
+        StateBTreeSetIter {
+            length:            self.len.try_into().unwrap_abort(),
+            next_node:         self.root,
+            depth_first_stack: Vec::new(),
+            tree:              self,
+            _marker_lifetime:  Default::default(),
+        }
+    }
+
+    /// Clears the set, removing all elements.
+    pub fn clear(&mut self)
+    where
+        S: HasStateApi, {
+        // Reset the information.
+        self.root = None;
+        self.next_node_id = NodeId {
+            id: 0,
+        };
+        self.len = 0;
+        // Then delete every node store in the state.
+        // Unwrapping is safe when only using the high-level API.
+        self.state_api.delete_prefix(&self.prefix).unwrap_abort();
+    }
+
+    /// Returns the smallest key in the set, which is strictly larger than the
+    /// provided key. `None` meaning no such key is present in the set.
+    pub fn higher(&self, key: &K) -> Option<StateRef<K>>
+    where
+        S: HasStateApi,
+        K: Serialize + Ord, {
+        let Some(root_node_id) = self.root else {
+            return None;
+        };
+
+        let mut node = self.get_node(root_node_id);
+        let mut higher_so_far = None;
+        loop {
+            let higher_key_index = match node.keys.binary_search(key) {
+                Ok(index) => index + 1,
+                Err(index) => index,
+            };
+
+            if node.is_leaf() {
+                return if higher_key_index < node.keys.len() {
+                    // This does not mutate the node in the end, just the representation in memory
+                    // which is freed after the call.
+                    Some(StateRef::new(node.keys.swap_remove(higher_key_index)))
+                } else {
+                    higher_so_far
+                };
+            } else {
+                if higher_key_index < node.keys.len() {
+                    // This does not mutate the node in the end, just the representation in memory
+                    // which is freed after the call.
+                    higher_so_far = Some(StateRef::new(node.keys.swap_remove(higher_key_index)))
+                }
+
+                let child_node_id = node.children[higher_key_index];
+                node = self.get_node(child_node_id);
+            }
+        }
+    }
+
+    /// Returns the largest key in the set, which is strictly smaller than the
+    /// provided key. `None` meaning no such key is present in the set.
+    pub fn lower(&self, key: &K) -> Option<StateRef<K>>
+    where
+        S: HasStateApi,
+        K: Serialize + Ord, {
```

Comment:
For this function we use `Ord` doing the binary search of the keys, it is probably fine to do a linear search instead.
But I am not sure how we would implement `insert` with only `PartialOrd`.

Reply:
On a normal architecture it would probably make sense to do a linear search for small collections, but for our case that's not going to be helpful I think since costs are assigned based on number of operations.